cohttp-eio/src/server.ml

```diff
@@ -1,205 +1,119 @@
-  let+ meth = Rwer.(token <* space) in
-  Http.Method.of_string meth
-
-let resource =
-  let open Eio.Buf_read.Syntax in
-  Rwer.(take_while1 (fun c -> c != ' ') <* space)
-
-let[@warning "-3"] http_request t =
-  let open Eio.Buf_read.Syntax in
-  let meth = meth t in
-  let resource = resource t in
-  let version = Rwer.(version <* crlf) t in
-  let headers = Rwer.http_headers t in
-  let encoding = Http.Header.get_transfer_encoding headers in
-  { Http.Request.meth; resource; version; headers; scheme = None; encoding }
-
-(* main *)
-
-let rec handle_request clock client_addr reader writer flow handler =
-  match http_request reader with
-  | request ->
-      let response, body = handler (request, reader, client_addr) in
-      write_response ~request clock writer (response, body);
-      if Http.Request.is_keep_alive request then
-        handle_request clock client_addr reader writer flow handler
-  | (exception End_of_file)
-  | (exception Eio.Io (Eio.Net.E (Connection_reset _), _)) ->
-      ()
-  | exception (Failure _ as ex) ->
-      write_response clock writer bad_request_response;
-      raise ex
-  | exception ex ->
-      write_response clock writer internal_server_error_response;
-      raise ex
-
-let connection_handler (handler : handler) clock flow client_addr =
-  let reader = Buf_read.of_flow ~initial_size:0x1000 ~max_size:max_int flow in
-  Buf_write.with_flow flow (fun writer ->
-      handle_request clock client_addr reader writer flow handler)
-
-let run_domain env ssock handler =
-  let on_error exn =
-    Printf.fprintf stderr "Error handling connection: %s\n%!"
-      (Printexc.to_string exn)
-  in
-  let handler = connection_handler handler env#clock in
-  Switch.run (fun sw ->
-      let rec loop () =
-        Eio.Net.accept_fork ~sw ssock ~on_error handler;
-        loop ()
+  handle ()
+
+let run socket server =
+  Eio.Switch.run @@ fun sw ->
+  let rec accept () =
+    let () =
+      Eio.Net.accept_fork ~sw socket ~on_error:raise
+      @@ fun socket peer_address ->
+      let () =
+        Logs.info (fun m ->
+            m "%a: accept connection" Eio.Net.Sockaddr.pp peer_address)
       in
-      loop ())
-
-let run ?(socket_backlog = 128) ?(domains = 1) ~port env handler =
-  Switch.run @@ fun sw ->
-  let domain_mgr = Eio.Stdenv.domain_mgr env in
-  let ssock =
-    Eio.Net.listen (Eio.Stdenv.net env) ~sw ~reuse_addr:true ~reuse_port:true
-      ~backlog:socket_backlog
-      (`Tcp (Eio.Net.Ipaddr.V4.loopback, port))
+      let input = Eio.Buf_read.of_flow ~max_size:max_int socket in
+      Eio.Buf_write.with_flow socket @@ fun output ->
+      callback server socket input output
+    in
+    accept ()
   in
-  for _ = 2 to domains do
-    Eio.Std.Fiber.fork ~sw (fun () ->
-        Eio.Domain_manager.run domain_mgr (fun () ->
-            run_domain env ssock handler))
-  done;
-  run_domain env ssock handler
-
-(* Basic handlers *)
-
-let not_found_handler _ = not_found_response
+  accept ()
```

Comment:
Could be just this, I think:
```ocaml
let run socket server =
  Eio.Net.run_server socket ~on_error:raise (fun socket peer_address ->
      Logs.info (fun m ->
          m "%a: accept connection" Eio.Net.Sockaddr.pp peer_address);
      let input = Eio.Buf_read.of_flow ~max_size:max_int socket in
      Eio.Buf_write.with_flow socket @@ fun output ->
      callback server socket input output
    )
```
But would probably be better to expose the connection handler function and let users call `Eio.Net.run_server` themselves, as it has extra arguments, and they can decide what to do about errors.